popovers: Enabled link result pastable both as a plain URL and HTML.

OmarAmeen01 opened the pull request:
## Made the feature to enable links paste able as both   URL  and   HTML on sites supporting HTML  

Fixes:  #31813 by @alya 
This feature  makes  links prettier on sites supporting HTML elements eg on google docs , emails and zulip etc as well  enable them to be paste able as links

- stream names are pasted as   `#streamName`
- topic names are pasted as `#streamName > topicName`

### Tests 
 **Tested the feature on** 
- chrome desktop (version : 129)
- firefox desktop ( version : 121)
## Description 

This is made with help of `native browser clipboard API `which replaces the` clipboardJs` as used by  the project before for copying the link 
with browsers `clipboard API ` we are able to copy text as both as` URL and HTML `that enables our links to be pretty when pasting them on sites supporting HTML  which gives a descriptive info about the link as well keeps its functionality intact that is redirecting user to that channel or topic
## demo 

https://github.com/user-attachments/assets/f569fd7a-8498-4959-b79b-b5124a943a83




<details>
- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented on web/src/topic_popover.js:
```diff
@@ -157,13 +157,22 @@ export function initialize() {
+                    const $topicLinkElement = $(".sidebar-popover-copy-link-to-topic");
+                    const formatedLinkText = `
+                    <a href=${$topicLinkElement.data("clipboard-text")}>${try_stream_topic_syntax_text($topicLinkElement.data("clipboard-text")).replaceAll(/\**/g, "\u200B")}</a>
+                   `;
```
Please read https://zulip.readthedocs.io/en/stable/subsystems/html-css.html and apply it to your code; we don't use camelCase, etc.

timabbott commented on web/src/stream_popover.js:
```diff
@@ -212,9 +213,21 @@ function build_stream_popover(opts) {
+                <a href=${$streamLinkElement.data("clipboard-text")}>${try_stream_topic_syntax_text($streamLinkElement.data("clipboard-text")).replaceAll(/\**/g, "\u200B")}</a>
+               `;
+                const clipboardItem = new ClipboardItem({
+                    "text/plain": new Blob([$streamLinkElement.data("clipboard-text")], {
```
Rather than reading `$streamLinkElement.data("clipboard-text")` from the DOM multiple times, just store it in a variable and use it in both places.

timabbott commented:
@kuv2707 do you have time to manually test and review this one?

kuv2707 commented on web/src/clipboard_handler.js:
```diff
@@ -0,0 +1,24 @@
+
+export function clipboard_handler($link_element, hide_popover, instance) {
+    const stream_topic_link_element_data = $link_element.data("clipboard-text");
+    const stream_topic_syntax_text = try_stream_topic_syntax_text(stream_topic_link_element_data);
```
This variable can be `null` if the stream or topic name contains some problematic characters (see #30071)

An example:

![image](https://github.com/user-attachments/assets/2a30c11a-211c-4c97-8608-704fc7ef5172)

kuv2707 replied:
Also this file should be `.ts`

kuv2707 reviewed (commented):
Left some comments. Feel free to tag me once you've addressed these.

OmarAmeen01 commented:
@kuv2707 sorry for disturbing I need to confirm this   do you want me to use  the argument that calls the function like this
in the arrow function 
 ` clipboard_handler( $(".copy_stream_link"),()=>popover_menus.hide_current_popover_if_visible(instance) );`       

 or should I make a functon like this 

` function hide_popover(instance){popover_menus.hide_current_popover_if_visible(instance)}`

 then use it like this

` clipboard_handler( $(".copy_stream_link"), hide_popover );`

kuv2707 commented:
The former. That will allow you to avoid passing `instance` to the function `clipboard_handler`, and you'd only need to call the callback after writing to the clipboard.

OmarAmeen01 commented:
@kuv2707  while addressing problem `This variable can be null if the stream or topic name contains some problematic characters `  I found the solution to the problem but I have to write different logic for` topic link and stream link` which  makes it pointless to put logic in separate file `clipboard_handler ` as they don't share common logic there are some parts common then I was thinking of using conditionals separatre topic link logic from stream link logic but it would cause repetition of code in the same file. what do you suggest ?

kuv2707 commented:
@OmarAmeen01 You can just check if `stream_topic_syntax_text` is null, and if so, return early from the function.

OmarAmeen01 commented:
@kuv2707  @timabbott  I used the approach you suggested  using this the link for the  stream or topic containing the **markdown symbols**  like you mentioned  in the **screenshot** won't be copied if that is okay I can raise the PR covering the edge  case of `null` only which will make that error go away .  If not then I have a ready made solution which fixes copying of topics  and stream links  containing the markdown syntax as well . Final call is yours I comfortable with any approach you suggest

kuv2707 commented:
You need to edit that commit to cover the `null` case, not open a new pr.

OmarAmeen01 commented:
sorry  That what I ment . editing commit to cover null

OmarAmeen01 commented on web/src/clipboard_handler.ts:
```diff
@@ -0,0 +1,35 @@
+export function clipboard_handler($link_element:JQuery, hide_popover:()=>void):void {
+
+    if(typeof $link_element.data("clipboard-text")==="string"){
+        // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
```
had to disable eslint on next line because next line returns type any by default it is built in Jquery

kuv2707 replied:
You can instead use `assert`s here. Look at other areas in the codebase where it is used.
In fact, there is no need to pass the link element here if all we do is extract one string from its dataset. We can directly pass that string here.

OmarAmeen01 commented:
@kuv2707  I have addressed the feedback you gave and amended the commit please take a look and give feedback

kuv2707 commented on web/src/clipboard_handler.ts:
```diff
@@ -0,0 +1,45 @@
+                });
+            return;
+        }
+        // The try_stream_topic_syntax_text() function returns the text
+        // as  #**stream_name > topic_name** to remove those two stars
+        // on fields supporting html in the start and end  we use
+        // replaceAll method that replaces those stars with html entity
+        // that is invisible however markdown still works on the zulip chat
```
This comment block is misplaced

kuv2707 replied:
Let's just remove this entirely.

OmarAmeen01 replied:
I used linter for it it is placed using built in linter And it was not there tim Abbot told me to leave a comment here so developer can know what's  going on here

kuv2707 replied:
The line of code that comment talks about is few lines above the comment.
And in any case it is too verbose for what it is describing

kuv2707 commented on web/src/clipboard_handler.ts:
```diff
@@ -0,0 +1,45 @@
+            stream_topic_link_element_data,
+        );
+
+        // invalid links or stream topic names containing markdown cause this variable to be null
```
I don't think this comment is required

kuv2707 commented on web/src/topic_popover.js:
```diff
@@ -157,13 +157,12 @@ export function initialize() {
-                        popover_menus.hide_current_popover_if_visible(instance);
-                    },
-                );
+                // copy link to the stream topic
```
Since this comment wasn't there earlier, lets not include it now.

kuv2707 commented on web/src/clipboard_handler.ts:
```diff
@@ -0,0 +1,45 @@
+
+        // invalid links or stream topic names containing markdown cause this variable to be null
+        if (stream_topic_syntax_text === null) {
+            navigator.clipboard
+                .writeText(stream_topic_link_element_data)
+                .then(hide_popover)
+                .catch(() => {
+                    blueslip.error("Clipboard access denied: requires user interaction.");
+                });
```
The variable becomes null if the link is invalid. We wouldn't wish to copy an invalid link to clipboard. Let us use assert here too, since we expect it not to be null ever.

kuv2707 commented on web/src/clipboard_handler.ts:
```diff
@@ -0,0 +1,45 @@
+import * as blueslip from "./blueslip";
+import {try_stream_topic_syntax_text} from "./copy_and_paste";
+
+export function clipboard_handler($link_element: JQuery, hide_popover: () => void): void {
```
```suggestion
export function clipboard_handler($link_element: JQuery, after_copy_cb: () => void): void {
```

kuv2707 reviewed (commented):
Left some comments

OmarAmeen01 commented:
@kuv2707  please have a look at this 

Uploading hightlight issue related to link pastable.mp4…

OmarAmeen01 commented:
@kuv2707  I have addressed your feedback and have made major changes to the `clipboard_handler.js` file here is the explanation why I did that please have a look  

https://github.com/user-attachments/assets/d3eeac30-7b48-4781-b71d-9a7c523f3fd6

kuv2707 commented on web/src/clipboard_handler.ts:
```diff
@@ -0,0 +1,36 @@
+    });
+    navigator.clipboard
+        .write([clipboardItem])
+        .then(after_copy_cb)
```
The `catch` callback will execute even when there is an error in the callback rather than writing to the clipboard.

kuv2707 commented on web/src/clipboard_handler.ts:
```diff
@@ -0,0 +1,36 @@
+import * as blueslip from "./blueslip";
+import {decode_stream_topic_from_url} from "./hash_util";
+import {get_sub_by_id} from "./stream_data";
```
Following the general pattern:
```suggestion
import * as stream_data from "./stream_data";
```

kuv2707 commented on tools/test-js-with-node:
```diff
@@ -67,6 +67,7 @@ EXEMPT_FILES = make_set(
         "web/src/browser_history.ts",
         "web/src/buddy_list.ts",
         "web/src/click_handlers.js",
+        "web/src/clipboard_handler.ts",
```
We should add some tests for the logic rather than exempting it altogether. You can refer to the docs for writing node tests.

kuv2707 commented on web/src/topic_popover.js:
```diff
@@ -1,9 +1,9 @@
 import render_delete_topic_modal from "../templates/confirm_dialog/confirm_delete_topic.hbs";
 import render_left_sidebar_topic_actions_popover from "../templates/popovers/left_sidebar/left_sidebar_topic_actions_popover.hbs";
 
+import {clipboard_handler} from "./clipboard_handler";
```
Here too

```suggestion
import * as clipboard_handler from "./clipboard_handler";
```

kuv2707 commented on web/src/clipboard_handler.ts:
```diff
@@ -0,0 +1,36 @@
+import {decode_stream_topic_from_url} from "./hash_util";
+import {get_sub_by_id} from "./stream_data";
+
+export function clipboard_handler(link_text: string, after_copy_cb: () => void): void {
```
We can rename this to `copy_to_clipboard`.

kuv2707 commented on web/src/clipboard_handler.ts:
```diff
@@ -0,0 +1,36 @@
+
+    if (!stream_topic) {
+        return;
+    }
```
Insert a newline here

kuv2707 commented on web/src/clipboard_handler.ts:
```diff
@@ -0,0 +1,36 @@
+        formatted_url = `
+        <a href="${link_text}">#${stream_name}</a>
+       `;
+    }
```
Insert a newline here

kuv2707 reviewed (commented):
Looks mostly good. Left some minor remarks.
It just needs some tests now.

OmarAmeen01 commented:
@kuv2707 @timabbott   I have added test for the core logic of the feature and also did some changes to the clipboard_handler.ts file  for making testing easy and giving each function a single responsibility please have look

timabbott commented on web/src/clipboard_handler.ts:
```diff
@@ -0,0 +1,42 @@
+    const stream_topic = hash_util.decode_stream_topic_from_url(link_text);
+
+    if (!stream_topic) {
+        return "Invalid stream";
```
I think `null` would be more appropriate for this.

timabbott commented on web/src/clipboard_handler.ts:
```diff
@@ -0,0 +1,42 @@
+import * as stream_data from "./stream_data";
+
+export function generate_formatted_link(link_text: string): string {
+    const stream_topic = hash_util.decode_stream_topic_from_url(link_text);
```
This should probably be called `link_data`, since a `message_id` can be in that object too.

timabbott commented on web/src/clipboard_handler.ts:
```diff
@@ -0,0 +1,42 @@
+    // render that name and treats that name as tag.
+
+    const topic_name = stream_topic.topic_name?.replace(/</g, "&lt;")?.replace(/>/g, "&gt;");
+    const stream_name = stream?.name?.replace(/</g, "&lt;")?.replace(/>/g, "&gt;");
```
Could this share code with or just use `get_fallback_markdown_link`?

timabbott commented:
I tested and pasting into a few places worked beautifully:
- Gmail
- Dropbox Paper

(In both cases you get a nicely labeled linked).

There's a couple things that I expected to work in this PR that don't seem to:
- Pasting into a GitHub issue/comment seems to just show the URL, i.e. use the plain-text version. 
- Pasting into the Zulip compose box seems to just show the URL, i.e. use the plain-text version. 

I'm a bit puzzled by the last one, because if I paste the content of https://github.com/zulip/zulip/pull/32018#pullrequestreview-2397450771 into chat.zulip.org, it seems to work, but not into the development environment, so maybe there's some regression there ... but it also seems plausible that this has a bug in not properly enclosing the link or something.

Certainly I'd have expected this to result in the Markdown link syntax, not just a URL:
```
    turndownService.addRule("links", {                                                                  
        filter: ["a"],                                                                                  
        replacement(content, node) {                                                                    
            assert(node instanceof HTMLAnchorElement);                                                  
            if (node.href === content) {                                                                
                // Checks for raw links without custom text.                                            
                return content;                                                                         
            }                                                                                           
            if (node.childNodes.length === 1 && node.firstChild!.nodeName === "IMG") {                  
                // ignore link's url if it only has an image                                            
                return content;                                                                         
            }                                                                                           
            return "[" + content + "](" + node.href + ")";                                              
        },                                                                                              
    });                                                                                                 
```

In any case, we should all presumably make this feature work with the "Copy link to message" feature in the message action menu.

timabbott commented:
I think the bug might be that `isUrl(trimmed_paste_text)` is passing here, so this logic is causing us to skip using the nicer HTML version. 
```
        // Only intervene to generate formatted links when dealing                                      
        // with a URL and a URL-safe range selection.                                                   
        if (isUrl(trimmed_paste_text)) {                                                                
            if (is_safe_url_paste_target($textarea)) {                                                  
                event.preventDefault();                                                                 
                event.stopPropagation();                                                                
                const url = trimmed_paste_text;                                                         
                compose_ui.format_text($textarea, "linked", url);                                       
                return;                                                                                 
            }                                                                                           
                                                                                                        
            if (                                                                                        
                !compose_ui.cursor_inside_code_block($textarea) &&                                      
                !cursor_at_markdown_link_marker($textarea) &&                                           
                !compose_ui.shift_pressed                                                               
            ) {                                                                                         
                // Try to transform the url to #**stream>topic** syntax                                 
                // if it is a valid url.                                                                
                const syntax_text = try_stream_topic_syntax_text(trimmed_paste_text);                   
                if (syntax_text) {                                                                      
                    event.preventDefault();                                                             
                    event.stopPropagation();                                                            
                    // To ensure you can get the actual pasted URL back via the browser                 
                    // undo feature, we first paste the URL in, then select it, and then                
                    // replace it with the nicer markdown syntax.                                       
                    add_text_and_select(trimmed_paste_text, $textarea);                                 
                    compose_ui.insert_and_scroll_into_view(syntax_text + " ", $textarea);               
                }                                                                                       
                return;                                                                                 
            }                                                                                           
        }                                                                                               
```

OmarAmeen01 commented:
> I tested and pasting into a few places worked beautifully:
> 
> * Gmail
> * Dropbox Paper
> 
> (In both cases you get a nicely labeled linked).
> 
> There's a couple things that I expected to work in this PR that don't seem to:......

@timabbott Pasting into a GitHub issue/comment seems to just show the URL. Because Github doesn't accept the `text/html` format of the copied item It just prefers the plain version of it and If we wanna make this compatible then we will either have to lose the functionality of formatting the text on Google Docs or create a new bottom that copies link as markdown for the sites like Github.  As `ClipboardItem` only supports `text/plain` and `text/html` right now and there is no `text/markdown` format right now that is supported by the browser so as per my knowledge it is technically impossible right now to get link formatted both as markdown, HTML and also be able to paste it as a plain link at the same time. 

The second thing that you mentioned about pasting the production stream or topic link is not getting formatted in the developer environment or vice versa. That is because the formatting of the link is done by this function `try_stream_topic_syntax_text` which internally depends on `get_sub_by_id` problem with this function is that it is only able to return the name of the stream that is created in that organization and won't be able to recognize ID of external stream name which is causing this bug. You can try this in the production environment as well if you copy the link to stream or topic and change the ID of the stream in the link to a random number and paste it in the compose box it won't be able to format it because the stream of that ID doesn't exist in its memory and this function will return null which will prevent its fromating hence the link gets pasted as it is. However, formatting works very well when we copy a link to the stream or topic that belongs to the organization I;e the coping link to stream or topic in organizations Zulip and paste it in their compose box it will format it as expected.

The solution to this problem is that `try_stream_topic_syntax_text` should reduce its dependence on the `get_sub_by_id`  function and get stream name via a custom function or using any other predefined function that doesn't depend on `stream_id` to retrieve its name. This way this issue will be resolved but for that new issue has to be raised and I will be happy to fix that.


The third problem you mentioned I:e make it work for copy link to messages as well. This was not in the issue request I thought Implementing it would be beyond the scope of the issue but If you permit I can enable the same functionality for that as well.

timabbott commented:
> This was not in the issue request I thought Implementing it would be beyond the scope of the issue but If you permit I can enable the same functionality for that as well

Yes, please add it; you can do it in an extra commit.

> The solution to this problem is that try_stream_topic_syntax_text should reduce its dependence on the get_sub_by_id function and get stream name via a custom function or using any other predefined function that doesn't depend on stream_id to retrieve its name.

We should be able to just process the `text/html` version using `turndown` when pasting into a different Zulip server.

OmarAmeen01 commented:
> Yes, please add it; you can do it in an extra commit.
sure will get it done. 
> We should be able to just process the `text/html` version using `turndown` when pasting into a different Zulip server; I don't think it needs to be able to recognize the Zulip syntax at all, necessarily.

 you are right there @timabbott  `isUrl(trimmed_paste_text)`  is passing. because this feature copies the URL as well as HTML so obviously this check passes to prevent this we should include another condition i:e  `isUrl(trimmed_paste_text && plain_html.length===0)` so that links containing both formats I:e `text/plain and text/html` won't be processed by this logic and are handled by the default `turndown service`. However `copied link to stream or topic` on paste will now generate a markdown format as that is what the `turndown` service does. 

I tested the above. It works fabulously even for links that are not from same server but currently, it won't work with production stream or topic links pasted into the developer environment and vice versa as that logic in production doesn't copy the link as both HTML and URL, till the changes in the PR are implemented it can't be tested in production. 
In order to test this you can make a new organization in Zulip developer environment and in `chrome/flags` add a link to that organization under this flag `unsafely-treat-insecure-origin-as-secure` and enable it because clipboard API doesn't work on unsecured sites for security reasons. 

If you are satisfied with this and permit me to fix this  I will fix this along with the changes you suggested in the next commit.

DJMcNab commented:
Fwiw, I only originally requested this feature for pasting the links to GitHub. I'd happily enable a setting to show a second button for that purpose only.

Of note, the `text/x-gfm` mime type is supported by GitHub. See e.g. https://jsfiddle.net/r3zsgha7/1/, which correctly pastes `GFM: [example.net](https://example.net)` for me in GitHub.
I don't know if that can be applicable in this case.

I don't know what condition is failing on the GitHub side to make pasting not work, because copying the rendered form of [Example link](https://example.com) works fine: `[Example link](https://example.com/)`. I presume there must be an over-zealous security check?

OmarAmeen01 commented:
@timabbott I have addressed your feedback and implemented the same for message links as well please have a look

OmarAmeen01 commented:
@DJMcNab  thank you for the suggestion. I tried implementing it but the function is then needed to be called without arguments  I:e it should be executed by copy functionalality of window. I have tested it extensively I will still check it can be integrated into the existing code.

kuv2707 commented:
@OmarAmeen01 Looks like [some feedback](https://github.com/zulip/zulip/pull/32018#discussion_r1830241502) is still unaddressed.

NotVishesh commented:
@OmarAmeen01 are you working on it, or can i continue your work?

timabbott commented:
Closing in favor of https://github.com/zulip/zulip/pull/32834, which continues this work.
